kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/expressions/DefaultExpressionEvaluator.java

```diff
@@ -0,0 +1,337 @@
+                    format("Column `%s` doesn't exist in input data schema: %s",
+                        column.getName(), inputDataSchema));
+            }
+            return new Tuple2<>(column, inputDataSchema.at(ordinal).getDataType());
+        }
+
+        @Override
+        Tuple2<Expression, DataType> visitCast(ImplicitCastExpression cast) {
+            throw new UnsupportedOperationException("CAST expression is not expected.");
+        }
+
+        private Predicate validateIsPredicateExpression(
+            Expression baseExpression,
+            Tuple2<Expression, DataType> expressionToValidate) {
+            checkArgument(
+                expressionToValidate._2 instanceof BooleanType,
+                "%s: expected a boolean type input expression, but got %s",
+                expressionToValidate._1,
+                expressionToValidate._2);
+            checkArgument(
+                expressionToValidate._1 instanceof Predicate,
+                "%s: expected a predicate type expression as input, but got %s",
+                baseExpression,
+                expressionToValidate._2);
+            return (Predicate) expressionToValidate._1;
+        }
+
+        private Expression validateAndRewriteBinaryComparator(Predicate predicate) {
+            checkArgument(predicate.getChildren().size() == 2, "expected two inputs");
+            Tuple2<Expression, DataType> leftResult = visit(predicate.getChildren().get(0));
+            Tuple2<Expression, DataType> rightResult = visit(predicate.getChildren().get(1));
+            Expression left = leftResult._1;
+            Expression right = rightResult._1;
+            DataType leftType = leftResult._2;
+            DataType rightType = rightResult._2;
+            if (!leftType.equivalent(rightType)) {
+                if (canCastTo(leftType, rightType)) {
+                    left = createCastExpression(left, rightType);
+                } else if (canCastTo(rightType, leftType)) {
+                    right = createCastExpression(right, leftType);
+                } else {
+                    String msg = format("%s: operands are of different types which are not " +
+                            "comparable: left type=%s, right type=%",
+                        predicate, leftType, rightType);
+                    throw new UnsupportedOperationException(msg);
+                }
+            }
+            return new Predicate(predicate.getName(), Arrays.asList(left, right));
+        }
+    }
+
+    /**
+     * Implementation of {@link ExpressionVisitor} to evaluate it on a {@link ColumnarBatch}.
+     */
+    private static class ExpressionEvaluator extends ExpressionVisitor<ColumnVector> {
+        private final ColumnarBatch input;
+
+        ExpressionEvaluator(ColumnarBatch input) {
+            this.input = input;
+        }
+
+        @Override
+        ColumnVector visitAnd(And and) {
+            Triplet<Integer, ColumnVector, ColumnVector> argResults =
```

Comment:
instead of using `Triplet<Integer, ColumnVector, ColumnVector>` in 4 places, its better to define a class here. Using `._2` is always harder to read that `.vector()`

Reply:
Defined a custom class.